pkg/streamingpromql/functions.go

```diff
@@ -141,6 +141,55 @@ func scalarToInstantVectorOperatorFactory(args []types.Operator, _ *limiting.Mem
 	return operators.NewScalarToInstantVector(inner, expressionPosition), nil
 }
 
+func LabelReplaceFunctionOperatorFactory() InstantVectorFunctionOperatorFactory {
+	return func(args []types.Operator, memoryConsumptionTracker *limiting.MemoryConsumptionTracker, _ *annotations.Annotations, expressionPosition posrange.PositionRange) (types.InstantVectorOperator, error) {
+		if len(args) != 5 {
+			// Should be caught by the PromQL parser, but we check here for safety.
+			return nil, fmt.Errorf("expected exactly 5 argument for label_replace, got %v", len(args))
+		}
+
+		inner, ok := args[0].(types.InstantVectorOperator)
+		if !ok {
+			// Should be caught by the PromQL parser, but we check here for safety.
+			return nil, fmt.Errorf("expected a range vector for 1st argument for label_replace, got %T", args[0])
+		}
+
+		dstLabel, ok := args[1].(types.StringOperator)
+		if !ok {
+			// Should be caught by the PromQL parser, but we check here for safety.
+			return nil, fmt.Errorf("expected a string for 2nd argument for label_replace, got %T", args[1])
+		}
+
+		replacement, ok := args[2].(types.StringOperator)
+		if !ok {
+			// Should be caught by the PromQL parser, but we check here for safety.
+			return nil, fmt.Errorf("expected a string for 3rd argument for label_replace, got %T", args[2])
+		}
+
+		srcLabel, ok := args[3].(types.StringOperator)
+		if !ok {
+			// Should be caught by the PromQL parser, but we check here for safety.
+			return nil, fmt.Errorf("expected a string for 4th argument for label_replace, got %T", args[3])
+		}
+
+		regex, ok := args[4].(types.StringOperator)
+		if !ok {
+			// Should be caught by the PromQL parser, but we check here for safety.
+			return nil, fmt.Errorf("expected a string for 5th argument for label_replace, got %T", args[4])
+		}
+
+		f := functions.FunctionOverInstantVector{
+			SeriesDataFunc:           functions.PassthroughData,
+			SeriesMetadataFunc:       functions.LabelReplaceFactory(dstLabel, replacement, srcLabel, regex),
+			NeedsSeriesDeduplication: true,
+		}
+
+		var o types.InstantVectorOperator = operators.NewFunctionOverInstantVector(inner, memoryConsumptionTracker, f, expressionPosition)
```

Comment:
[nit]

```suggestion
		o := operators.NewFunctionOverInstantVector(inner, memoryConsumptionTracker, f, expressionPosition)
```

Reply:
This is needed because `NewFunctionOverInstantVector` returns `*FunctionOverInstantVector`

Reply:
Not sure I follow sorry - that sounds like it shouldn't be an issue, and making the change locally seems to work fine.

Reply:
really? It makes line 55 below incompatible:

``grafana/mimir/pkg/streamingpromql/functions.go:55:8: cannot use operators.NewDeduplicateAndMerge(o, memoryConsumptionTracker) (value of type *operators.DeduplicateAndMerge) as *operators.FunctionOverInstantVector value in assignment
```

Reply:
I think you're looking at the wrong method, line 55 is `SingleInputVectorFunctionOperatorFactory` :) But I'll approve this to unblock you.

Reply:
Whoops, thanks, I absolutely was. ETOOMANYFILETABSOPEN. Since it's still waiting to merge I pushed it onto here.